corehq/util/tests/test_couch.py

```diff
@@ -76,3 +79,152 @@ def test_get_document_or_404_success(self):
+            for group in self.groups:
+                group['test_property'] = True
+                iter_db.save(group)
+        for group in self.groups:
+            new = self.db.get(group._id)
+            self.assertTrue(new.get('test_property', False))
+
+    def test_conflicted_doc(self):
+        old = self.groups[3]
+        new = Group.get(old._id)
+        new.name = "bwahahaha"
+        new.save()
+        self.assertEqual(old._id, new._id)
+        self.assertNotEqual(old._rev, new._rev)
+        # now if you try to save 'old', it should conflict
+        with IterDB(self.db, chunksize=5) as iter_db:
+            for group in self.groups:
+                iter_db.save(group)
+
+        should_succeed = {g._id for g in self.groups if g._id != old._id}
+        self.assertEqual(should_succeed, iter_db.saved_ids)
+
+        self.assertEqual({old._id}, iter_db.error_ids)
+
+    def test_delete(self):
+        with IterDB(self.db, chunksize=5) as iter_db:
+            deleted_groups = set()
+            for group in self.groups[:4]:
+                deleted_groups.add(group._id)
+                iter_db.delete(group)
+
+            saved_groups = set()
+            for group in self.groups[4:]:
+                saved_groups.add(group._id)
+                iter_db.save(group)
+
+        self.assertEqual(deleted_groups, iter_db.deleted_ids)
+        self.assertEqual(saved_groups, iter_db.saved_ids)
+        for group_id in deleted_groups:
+            self.assertRaises(ResourceNotFound, self.db.get, group_id)
+
+    def test_iter_update(self):
+        def desired_action(group):
+            i = group['index']
+            if i == 1:
+                return 'DELETE'
+            elif i % 2 == 0:
+                return 'UPDATE'
+            else:
+                return 'IGNORE'
+
+        def mark_cool(group):
+            action = desired_action(group)
+            if action == 'UPDATE':
+                group['is_cool'] = True
+                return DocUpdate(group)
+            elif action == 'DELETE':
+                return DocUpdate(group, delete=True)
+
+        ids = [g._id for g in self.groups] + ['NOT_REAL_ID']
+        res = iter_update(self.db, mark_cool, ids)
+        self.assertEqual(res.not_found_ids, {'NOT_REAL_ID'})
+        for result_ids, action in [
+            (res.ignored_ids, 'IGNORE'),
+            (res.deleted_ids, 'DELETE'),
+            (res.updated_ids, 'UPDATE'),
+        ]:
+            self.assertEqual(
+                result_ids,
+                {g._id for g in self.groups if desired_action(g) == action}
+            )
+
+    def test_iter_update_no_actual_changes(self):
+        ids = [g._id for g in self.groups]
+        # tell iter_update to save the docs, but don't actually make changes
+        res = iter_update(self.db, lambda doc: DocUpdate(doc), ids)
+        self.assertEqual(res.ignored_ids, set(ids))
+
+    def test_iter_update_bad_return(self):
```

Comment:
Now the transform function has to return either `None` or an instance of `DocUpdate`